src/web/page.rs

```diff
@@ -119,42 +111,145 @@ impl<T: ToJson> Page<T> {
-        );
-        tree.insert(
-            "cratesfyi_version".to_owned(),
-            crate::BUILD_VERSION.to_json(),
-        );
-        tree.insert(
-            "cratesfyi_version_safe".to_owned(),
-            crate::BUILD_VERSION
+        state.serialize_field("content", &self.content)?;
+        state.serialize_field("rustc_resource_suffix", self.rustc_resource_suffix)?;
+        state.serialize_field("cratesfyi_version", crate::BUILD_VERSION)?;
+        state.serialize_field(
+            "cratesfyi_version_safe",
+            &crate::BUILD_VERSION
                 .replace(" ", "-")
                 .replace("(", "")
-                .replace(")", "")
-                .to_json(),
-        );
-        tree.insert("varss".to_owned(), self.varss.to_json());
-        tree.insert("varsb".to_owned(), self.varsb.to_json());
-        tree.insert("varsi".to_owned(), self.varsi.to_json());
-        Json::Object(tree)
+                .replace(")", ""),
+        )?;
+        state.serialize_field("varss", &self.varss)?;
+        state.serialize_field("varsb", &self.varsb)?;
+        state.serialize_field("varsi", &self.varsi)?;
+
+        state.end()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::super::releases::{self, Release};
+    use super::*;
+    use serde_json::json;
+
+    #[test]
+    fn serialize_page() {
+        let time = time::get_time();
+
+        let mut release = Release::default();
+        release.name = "lasso".into();
+        release.version = "0.1.0".into();
+        release.release_time = time.clone();
+
+        let mut varss = BTreeMap::new();
+        varss.insert("test".into(), "works".into());
+        let mut varsb = BTreeMap::new();
+        varsb.insert("test2".into(), true);
+        let mut varsi = BTreeMap::new();
+        varsi.insert("test3".into(), 1337);
+
+        let page = Page {
+            title: None,
+            content: vec![release.clone()],
+            status: status::Status::Ok,
+            varss,
+            varsb,
+            varsi,
+            rustc_resource_suffix: &*RUSTC_RESOURCE_SUFFIX,
+        };
+
+        let correct_json = json!({
+            "content": [{
+                "name": "lasso",
+                "version": "0.1.0",
+                "description": null,
+                "target_name": null,
+                "rustdoc_status": false,
+                "release_time": super::super::duration_to_str(time),
+                "release_time_rfc3339": time::at(time).rfc3339().to_string(),
+                "stars": 0
+            }],
+            "varss": { "test": "works" },
+            "varsb": { "test2": true },
+            "varsi": { "test3": 1337 },
+            "rustc_resource_suffix": &*RUSTC_RESOURCE_SUFFIX,
+            "cratesfyi_version": crate::BUILD_VERSION,
+            "cratesfyi_version_safe": crate::BUILD_VERSION
+                .replace(" ", "-")
+                .replace("(", "")
+                .replace(")", ""),
```

Comment:
Can you add a test for this as an example? Maybe refactor it into a small function? I understand that `BUILD_VERSION` is not predictable from within the code, but you could hard-code the current version, `0.6.0 (44f6f29 2020-05-26)`.

Reply:
If there is a way to test whether the generated `BUILD_VERSION` is a valid URL (maybe using `Url::parse`?) that would also be great, since we use `build_version_safe` in several query parameters.